### PR TITLE
[Gatsby] New Code Page

### PR DESCRIPTION
**Pull request by ardatan**
On this PR, We started by improving the design and ordering of the code page   in addition to our Gatsby migration PR https://github.com/graphql/graphql.github.io/pull/938.
We’ve followed the guidelines and ideas from https://github.com/graphql/graphql.github.io/issues/768.
There are now clear sections for each language.
Also, this PR uses a ranking algorithm for libraries and tools in the Code page. 
This algorithm compares those by their GitHub stars, NPM downloads, number of commits (except bots like renovate and dependabot) in the last month. 

We deployed the new Code page;
http://guild-gatsby-graphql-org-new-code-page.netlify.app/code

This PR includes changes from;
https://github.com/graphql/graphql.github.io/pull/954
https://github.com/graphql/graphql.github.io/pull/927
https://github.com/graphql/graphql.github.io/pull/774
https://github.com/graphql/graphql.github.io/pull/673
https://github.com/graphql/graphql.github.io/pull/664
https://github.com/graphql/graphql.github.io/pull/916
https://github.com/graphql/graphql.github.io/pull/914
https://github.com/graphql/graphql.github.io/pull/956
https://github.com/graphql/graphql.github.io/pull/950
https://github.com/graphql/graphql.github.io/pull/922
https://github.com/graphql/graphql.github.io/pull/903
https://github.com/graphql/graphql.github.io/pull/859
https://github.com/graphql/graphql.github.io/pull/858
https://github.com/graphql/graphql.github.io/pull/865
https://github.com/graphql/graphql.github.io/pull/781

### CODE REVIEW

**Comment by carolstran** [comment]
There's a lot happening here, but an initial question: Why have the changes to the Code page in the same PR as the migration?

Asking because the Code page updates will require more review and is more... controversial... than the migration (which has already been approved). Plus they are two separate issues which would traditionally mean two PRs, so I'm curious about the thinking behind it.

**Comment by Urigo** [comment]
Thank you for the comment @carolstran 
I would welcome any comments about the code page and if we'll get comments that would make it controversial or hold us back from merging, we'll pull those changes into a separate PR.

**Comment by IvanGoncharov** [comment]
@Urigo I agree with @carolstran  that we need to separate migration from design and content changes.
Can you please split code page changes into separate PR?

@ardatan can you please properly rebase this PR on top of `master`?

**Comment by ardatan** [comment]
Due to naming of the branch, this PR has code page redesign. And the gatsby migration without code page redesign is on another PR;
https://github.com/graphql/graphql.github.io/pull/938

**Comment by carolstran** [comment]
I might wait until after #938 is merged to properly review (it's difficult to see what the specific Code page changes are now), but a quick question in the meantime: Did you have a chance to check out the discussion in #911? If so, I'd be curious to hear how it informed the way you implemented a ranking algorithm.

**Comment by Urigo** [comment]
Thanks for the review @carolstran, I worked with @ardatan on the algorithm implementation so I'll try to explain our reasoning behind it.

We’ve basically followed the comment you quoted at the beginning of the issue:

The ranking is calculated by Github stars, Commit volume and Latest commit which are not renovate/bot and package downloads stats.
We believe this follows the main goal better than the current state of rank on the website.

The reasoning behind the algorithm and the goals are as follows:

We see the main goal of graphql.org is to make it as easy as possible for beginners to get into GraphQL without getting stuck.
That’s the first place beginners are being introduce to GraphQL.
More advanced developers would find their path with more confidence.

With the current state to rank, beginners are constantly trying to use outdated, unmaintained libraries and get stuck.
(this is feedback we get daily for months and years....)
This ranking puts the most popular, actively maintained libraries on top, to minimize the chances for a beginner to hit a path that doesn’t work.

The reason we wanted to get this merged quickly was that we agree with the sentiment of [that comment](https://github.com/graphql/graphql.github.io/issues/911#issuecomment-686924140) by @dosco - “Sometimes good enough is better than perfect.”
As this is already better, and also we are going to actively and continuously work on graphql.org from now on and improve it with any feedback that would come.
That algorithm would also make the job of merging a lot of community libraries much easier, as we don’t need to have a whole discussion on each of them (and would close tons of open issues and PRs), and we could continue the discussion about improving the algorithm separately.

We would gradually apply that algorithm and the parts of it where possible.
Where it won’t work or achieve its goal, lets fall back to alphabetical order.

Once that will be merged, we will add more package managers data from package managers that are not npm for other languages and ecosystems.

It’s similar also to what was suggested [on this comment](https://github.com/graphql/graphql.github.io/issues/911#issuecomment-687167462) by @jeggy

We could also later on add voting mechanism like @saihaj suggested [here](https://github.com/graphql/graphql.github.io/issues/911#issuecomment-688415811), even though I believe that could easily be manipulated and also it’s much more complicated because it means we need to start storing state and votes somewhere...

Sorry for the long message @carolstran, looking forward to hearing your thoughts and feedback!

**Review by Urigo — APPROVED** [review]

File: gatsby-node.js
Comment (orta) [inline gatsby-node.js:37]: This section again is continuing to get chunks of hard to grok code without the reasoning for why. I think this should probably have comments - and more likely I think that async could probably just be moved to a bootstrapping phrase.

**Review by orta — CHANGES_REQUESTED** [review]
I think this needs a fundamental re-think, it forces anyone wanting to edit the site to have `GITHUB_ACCESS_TOKEN` set up, and anyone wanting to add to a library to make markdown edits inside a JSON file. I don't want this repo to have these barriers to entry.

How I solved this in the TypeScript website was to have a `bootstrap` command which [bundles up a series of markdown](https://github.com/microsoft/TypeScript-Website/tree/v2/packages/tsconfig-reference/copy/en/options) files which are easy for folks to contribute to (and to add their own). 

You could have a folder structure like:

```
./src/content
├── blog
│   ├── 20150914-graphql.md
│   ├── 20151016-subscriptions.md
│   ├── 20160419-mocking.md
│   ├── 20160502-rest-api-graphql-wrapper.md
│   └── 20160914-production-ready.md
├── codeofconduct
│   └── CodeOfConduct.md
├── code
│   ├── JavaScript
│   │    ├── Apollo.md
│   │    └── Relay.md
│   ├── Go
│   │    └── GraphQL-Go.md
│   ├── Groovy
│   │    └── Gorm-GraphQL.md

```

The metadata can be attached as yml frontmatter.

Then during bootstrap the additional sorting metadata can be grabbed _if_ there's `GITHUB_ACCESS_TOKEN` but there should be some sort of fallback to make the site work if that's not available. That could either be a mock/fake, "old" data in source or just be optional in the UI.

**Comment by benjie** [comment]
Could I request the addition of [PostGraphile](https://github.com/graphile/postgraphile) to this? I think under the JavaScript section makes most sense.

Separately,

> This algorithm compares those by their GitHub stars, NPM downloads, number of commits (except bots like renovate and dependabot) in the last month.

I'd just like to point out that ranking based on number of commits in the last month unfairly penalises projects that use the squash-merge technique. Maybe LOC changed might be a fairer metric, though again that's problematic as some languages are naturally more terse than others. Personally I feel like this metric should be removed, or should be replaced with something like "time since last commit."

**Comment by orta** [comment]
Personally, I don't really see much user value in sorting each library per language if we provide better visibility into useful metrics. The page basically looks randomly ordered right now on this branch. I'm not sure sorting internally vs alphabetically it pays for itself, I think by most recent deploy could work, though we'd probably need to explain that somewhere on the page. Though it might be more obvious in the new designs. I think there's definitely value in sorting the languages by stars though.

Rather than a write-up, I've worked on the designs and added all my comments, the previous one was only meant to be a sketch to get a direction from folks and so I've just continued that direction and tightened it up.

[Figma link](https://www.figma.com/file/iM8ItJt4SrVU6bPUdZjpHm/GraphQL-Code-Page?node-id=0%3A1)

graphql-code-page

**Comment by Urigo** [comment]
@orta :
1. Good point about the barriers of entry - We'll fix that to make it easy for anyone to add entries - this new system should be easier and more structure and not harder... also thanks for the resources!
2. Also I love your design and points!!! - we'll work on it this week and try to set up a new version
3. I also agree sorting the languages as well makes sense for the user!   (sum of overall stars is good)
4. About language icons, I'm also undecided.  it makes it easier to spot for visitors. maybe we could support people who add new languages and if they don't have icons, help them find it instead of blocking their PRs?
5. Code example - should we allow every entry to have code example?

@benjie :
1. Of course we'll add PostGraphile in here! ❤️ (and all the currently open PRs for resources)
2. About the ranking - I personally think its really important and valuable (for beginners coming to find up to date resources  and for active maintainers of libraries in the ecosystem) so I would rather having us working together on a solution rather then completely removing it.

Our first iteration was "time since last commit."
It worked ok so if that's something you agree with, we are happy to fall back to it instead of completely removing it.

We changed it later because we saw some libraries that just do some update deps commits every once in a while without a bot and thought to maybe find a more robust measurement.
Maybe your suggestion to use LOC is good?  as this measure is for comparing libraries in the same language, it can be an ok one?
Again - it is obvious there is not perfect measure here and we can find 1000 ways to "trick" it, but I believe any solution from the mentioned above is better then removing it completely as if someone invest the time to "trick" it, at least they are spending time looking at their libraries :)

**Comment by benjie** [comment]
Thanks @Urigo! I understand the motivation behind this, but I think number of commits in X period, LOC, and time since last commit are all poor metrics for project activity - they incentivize a particular way of developing a project (e.g. monorepo, documentation in same repo as project, examples in same repo as project, etc) and penalize other approaches, and similarly the prioritise new projects over more mature projects that change more gradually. This kind of subtlety is really hard to capture in an automated fashion. I wonder if more of a boolean approach is appropriate: active (e.g. in the last 3 months) vs inactive, rather than "updated 4 hours ago" vs "updated 25 days ago" (which could both be equally active IMO). What do you think?

**Comment by Urigo** [comment]
@benjie good point, I like that.
Let's do active/not active according to 3 months (excluding bots)!

**Review by orta — APPROVED** [review]
I've given it a run through locally and shipped anything I had as feedback, I think this is a great improvement. 

It means we can accept pretty much any PR to the code section now without merge conflicts, and we can encourage people to show example code in their samples wiht as much code as the want. Good work.

I'll leave this open for a day for feedback, if there's none blocking. I'll merge.

**Comment by orta** [comment]
Alright, let's get this merged. Thanks folks!

**Pull request merged**
